crmsh/upgradeutil.py

```diff
@@ -0,0 +1,103 @@
+import logging
+import os.path
+import parallax
+import typing
+
+from . import parallax as crmparallax
+from . import remoteutil
+from . import utils as crmutils
+
+
+# pump this seq when upgrade check need to be run
+CURRENT_UPGRADE_SEQ = 1
+DATA_DIR = os.path.expanduser('~hacluster/crmsh')
+SEQ_FILE_PATH = DATA_DIR + '/upgrade_seq'
+FORCE_UPGRADE_FILE_PATH = DATA_DIR + '/upgrade_forced'
+
+
+logger = logging.getLogger(__name__)
+
+
+def _parallax_call(nodes, cmd):
+    parallax_options = parallax.Options()
+    parallax_options.ssh_options = ['StrictHostKeyChecking=no', 'ConnectTimeout=10']
+    return parallax.run(nodes, cmd, parallax_options)
+    # parallax.call raises exception if rc != 0, use an alternative implementation until it is fixed
+    #ret = dict()
+    #for node in nodes:
+    #    try:
+    #        result = remoteutil.run(node, cmd, use_shell=True)
+    #        result = (result.rc, result.stdout, result.stderr)
+    #    except remoteutil.Error as e:
+    #        logger.debug("Error calling %s on %s", cmd, node, exc_info=e)
+    #        result = e
+    #    ret[node] = result
+    #return ret
+
+
+def _upgrade_if_needed(nodes, fn: typing.Callable[[int], None]):
+    seq = 0
+    # FIXME: too slow, check localhost only
+    results = list(_parallax_call(nodes, 'test -f {}'.format(FORCE_UPGRADE_FILE_PATH)).values())
+    if any(isinstance(item, Exception) for item in results):
+        logger.info("parallax failed: nodes: %s, results: %s", nodes, results)
+        return
```

Comment:
It is a good point. A cluster should eventually be upgraded to a consistent version. So I would rather not to upgrade when some nodes are unreachable, unless the cluster is quite large and the possibility of temporary network failure were significantly high so that it were impossible to reach all nodes at the same time.

It also remains me that I should test more failing cases.